src/osd/scrubber/scrub_machine.h

```diff
@@ -776,22 +842,91 @@ struct ReplicaActive : sc::state<
       const ReplicaReserveReq&,
       bool async_request);
 
+  /**
+   * the queued reservation request was granted by the async reserver.
+   * Notify the Primary.
+   * Returns 'false' if the reservation is not the last one to be received
+   * by this replica.
+   */
+  bool granted_by_reserver(const AsyncScrubResData& resevation);
+
   /// handle a 'release' from a primary
   void on_release(const ReplicaRelease& ev);
 
-  /// cancel the reserver request.
-  /// The 'failure' re 'log_failure' is logged if we are not reserved to
-  /// begin with.
-  void clear_reservation_by_remote_primary(bool log_failure);
+  /**
+   * cancel a granted or pending reservation
+   *
+   * warn_if_no_reservation is set to true if the call is in response to a
+   * cancellation from the primary.  In that event, we *must* find a
+   * a granted or pending reservation and failing to do so warrants
+   * a warning to clog as it is a bug.
+   */
+  void clear_remote_reservation(bool warn_if_no_reservation);
+
+  /**
+   * discard (and log) unhandled 'reservation granted' messages
+   * from the async reserver.
+   * As canceled reservations may still be triggered, this is not
+   * necessarily a bug.
+   */
+  void ignore_unhandled_grant(const ReserverGranted&);
 
-  using reactions = mpl::list<sc::transition<IntervalChanged, NotActive>>;
+  using reactions = mpl::list<
+      sc::transition<IntervalChanged, NotActive>,
+      sc::in_state_reaction<
+	  ReserverGranted,
+	  ReplicaActive,
+	  &ReplicaActive::ignore_unhandled_grant>>;
 
  private:
-  bool reserved_by_my_primary{false};
-
-  // shortcuts:
   PG* m_pg;
   OSDService* m_osds;
+
+  // --- remote reservation machinery
+
+  /*
+   * 'reservation_granted' is set to 'true' when we have grant confirmation
+   *  to the primary, and the reservation has not yet been canceled (either
+   *  by the primary or following an interval change).
+   *
+   * Note the interaction with 'pending_reservation_nonce': the combination
+   * of these two variables is used to track the state of the reservation
+   * with the scrub_reserver. The possible combinations:
+   * - pending_reservation_nonce == 0 && !reservation_granted -- no reservation
+   *    was granted, and none is pending;
+   * - pending_reservation_nonce != 0 && !reservation_granted -- we have a
+   *   pending cb in the AsyncReserver for a request with nonce
+   *   'pending_reservation_nonce'
+   * - pending_reservation_nonce == 0 && reservation_granted -- we have sent
+   *   a response to the primary granting the reservation
```

Comment:
Let's state the invariant explicitly:

```
 * invariant: !((pending_reservation_nonce != 0) && reservation_granted)
```

It's also worth noting how these states relate to the synchronous/compat case:

```
 * Note that in the event that the primary is too old to support asynchronous reservation,
 * MOSDScrubReserve::wait_for_resources will be set to false by the decoder and we
 * bypass the 2nd case above.  See ReplicaActive::on_reserve_request.
```